subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c

```diff
@@ -133,6 +139,91 @@ img_mgmt_state_flags(int query_slot)
 }
 #endif
 
+/* Return number of slot that is considered candidate for next boot */
+int img_mgmt_get_opposite_slot(int slot)
+{
+	if (slot < 0 ||
+	    slot > (CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER) << 1) {
+		LOG_DBG("Impossible slot number: %d", slot);
+		return -1;
+	}
+
+	return (slot ^ 1);
+}
+
+#ifndef CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP
+int img_mgmt_get_next_boot_slot(int image, int *type)
+{
+	int state = mcuboot_swap_type_multi(image);
+	int slot = 0;		/* Default to slot 0 as next boot slot */
+	int lt = NEXT_BOOT_TYPE_NORMAL;
+
+	switch (state) {
+	case BOOT_SWAP_TYPE_NONE:
+			slot = image << 1;
+		break;
+	case BOOT_SWAP_TYPE_PERM:
+		/* Type is NEXT_BOOT_TYPE_NORMAL as it means that the returned slot
+		 * is boot slot is normal slot for all next boots.
+		 */
+		slot = (image << 1) + 1;
+		break;
+	case BOOT_SWAP_TYPE_REVERT:
+		{
+			const int active_slot = img_mgmt_active_slot(image);
+
+			slot = img_mgmt_get_opposite_slot(active_slot);
+			/* App has boot to be tested and has not yet been confirmed, which means
+			 * that next boot will be revert to reported slot.
+			 */
+			lt = NEXT_BOOT_TYPE_REVERT;
+		}
+		break;
+	case BOOT_SWAP_TYPE_TEST:
+		{
+			const int active_slot = img_mgmt_active_slot(image);
+
+			slot = img_mgmt_get_opposite_slot(active_slot);
+			/* Reported next boot slot is set for one boot only and app needs to
+			 * confirm itself or it will be reverted.
+			 */
+			lt = NEXT_BOOT_TYPE_TEST;
+		}
+		break;
+	default:
+		/* Should never, ever happen */
+		LOG_ERR("Unexpected swap state %d", state);
+		return -1;
+	}
+	LOG_DBG("(%d, *) => slot = %d, type = %d",
+		image, slot, lt);
+
+	if (type != NULL) {
+		*type = lt;
+	}
+	return slot;
+}
+#else
+int img_mgmt_get_next_boot_slot(int image, int *type)
+{
+	struct image_version aver;
+	struct image_version over;
+	int active_slot = img_mgmt_active_slot(image);
+	int other_slot = img_mgmt_get_opposite_slot(active_slot);
+	*type = NEXT_BOOT_TYPE_NORMAL;
+
+	int rcs = img_mgmt_read_info(other_slot, &over, NULL, NULL);
+	int rca = img_mgmt_read_info(active_slot, &aver, NULL, NULL);
+
+	if (rcs == 0 && rca == 0 && img_mgmt_vercmp(&aver, &over) < 0) {
```

Comment:
(likewise with this, I wonder how a direct XIP configured system would work on e.g. nRF53 where the secondary image would not be direct XIP as it would be loaded to the network core, maybe we need some new option in dts per partition which specifies what type of upgrade system it uses... seems like a messy idea)

Reply:
Currently we do not support DirectXIP with nRF53 as this would required mixed behaviour where app core image is DirectXIPed but the net core needs to be transferred to to the core, which means swapped.

I would actually prefer the nRF53 two core boots to desist using MCUboot for swapping net core image and instead drop that job on an application: this way MCUboot has to care about booting the app core image only and net core will be handled the same way whether swap or DirectXIP is used. Also, if net core image becomes part of app core image we have greater flexibility in flash partitioning, and less fragmentation, as we do not have to use special partitions for net core anymore.